File: backend/src/main/java/com/team48/inscriptionscolaire/notification/NotificationService.java
package com.team48.inscriptionscolaire.notification;

import com.team48.inscriptionscolaire.user.User;
import com.team48.inscriptionscolaire.user.UserRepository;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class NotificationService {
    private final SimpMessagingTemplate messagingTemplate;
    private final UserRepository userRepository;
    private final NotificationRepository notificationRepository;

    @Autowired
    public NotificationService(SimpMessagingTemplate messagingTemplate, 
                              UserRepository userRepository,
                              NotificationRepository notificationRepository) {
        this.messagingTemplate = messagingTemplate;
        this.userRepository = userRepository;
        this.notificationRepository = notificationRepository;
    }

    /*public void sendGlobalNotification() {
        Notification notification = new Notification("This is a global notification!");
        messagingTemplate.convertAndSend("/topic/notifications", notification);
    }*/

    /**
     * Envoie une notification privée à un utilisateur spécifique et la persiste en base de données.
     * @param userId L'identifiant de l'utilisateur (doit correspondre à son nom d'utilisateur de connexion)
     * @param message Le contenu du message à envoyer
     */
    @Transactional
    public void sendPrivateNotification(final String userId, final String message) {
        try {
            // Find the user by email/username
            Optional<User> userOptional = userRepository.findByEmail(userId);
            if (userOptional.isPresent()) {
                User user = userOptional.get();
                
                // Create and persist the notification
                Notification notification = new Notification();
                notification.setContent(message);
                notification.setUser(user);
                notification.setRead(false);
                notification = notificationRepository.save(notification);
                
                // Send via WebSocket
                messagingTemplate.convertAndSendToUser(userId, "/topic/private-notifications", notification);
            }
        } catch (Exception e) {
            // Log the error but don't throw exception to avoid breaking the enrollment process
            e.printStackTrace();
        }
    }
    
    /**
     * Envoie une notification à tous les administrateurs et les persiste en base de données.
     * @param message Le contenu du message à envoyer
     */
    @Transactional
    public void sendNotificationToAdmins(final String message) {
        try {
            List<User> adminUsers = userRepository.findAdminUsers();
            
            for (User admin : adminUsers) {
                try {
                    // Create and persist the notification for each admin
                    Notification notification = new Notification();
                    notification.setContent(message);
                    notification.setUser(admin);
                    notification.setRead(false);
                    notification = notificationRepository.save(notification);
                    
                    // Send via WebSocket
                    messagingTemplate.convertAndSendToUser(admin.getUsername(), "/topic/private-notifications", notification);
                } catch (Exception e) {
                    // Log the error but continue with other admins
                    e.printStackTrace();
                }
            }
        } catch (Exception e) {
            // Don't throw exception to avoid breaking the enrollment process
            e.printStackTrace();
        }
    }
    
    /**
     * Get all notifications for the current authenticated user
     */
    public List<Notification> getNotificationsForCurrentUser() {
        User currentUser = getCurrentUser();
        return notificationRepository.findByUserIdOrderByCreatedDateDesc(currentUser.getId());
    }
    
    /**
     * Get a specific notification by ID
     */
    public Notification getNotificationById(Integer id) {
        Optional<Notification> notification = notificationRepository.findById(id);
        return notification.orElse(null);
    }
    
    /**
     * Mark all notifications as read for the current user
     */
    @Transactional
    public void markAllAsReadForCurrentUser() {
        User currentUser = getCurrentUser();
        List<Notification> notifications = notificationRepository.findByUserIdAndIsReadFalse(currentUser.getId());
        
        for (Notification notification : notifications) {
            notification.setRead(true);
        }
        
        notificationRepository.saveAll(notifications);
    }
    
    /**
     * Delete a notification by ID
     */
    @Transactional
    public boolean deleteNotification(Integer id) {
        Optional<Notification> notification = notificationRepository.findById(id);
        
        if (notification.isPresent()) {
            notificationRepository.deleteById(id);
            return true;
        }
        
        return false;
    }
    
    /**
     * Get count of unread notifications for the current user
     */
    public long getUnreadNotificationsCountForCurrentUser() {
        User currentUser = getCurrentUser();
        return notificationRepository.countByUserIdAndIsReadFalse(currentUser.getId());
    }
    
    /**
     * Get the current authenticated user
     */
    private User getCurrentUser() {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }
}

File: frontReact/src/services/notificationService.js
import api from './api';

const NOTIFICATIONS_URL = '/api/v1/notifications'; // Base path for notification endpoints

/**
 * Service function to get all notifications for the current user.
 * @returns {Promise} A promise that resolves to the list of notifications.
 */
export const getAllNotifications = async () => {
  try {
    const response = await api.get(`${NOTIFICATIONS_URL}/all`);
    return response.data;
  } catch (error) {
    console.error('Error in getAllNotifications:', error);
    throw error;
  }
};

/**
 * Service function to get a specific notification by ID.
 * @param {number} id The ID of the notification to fetch.
 * @returns {Promise} A promise that resolves to the notification object.
 */
export const getNotificationById = async (id) => {
  try {
    const response = await api.get(`${NOTIFICATIONS_URL}/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error in getNotificationById with id ${id}:`, error);
    throw error;
  }
};

/**
 * Service function to mark all notifications as read for the current user.
 * @returns {Promise} A promise that resolves when all notifications are marked as read.
 */
export const markAllNotificationsAsRead = async () => {
  try {
    const response = await api.post(`${NOTIFICATIONS_URL}/mark-all-as-read`);
    return response.data;
  } catch (error) {
    console.error('Error in markAllNotificationsAsRead:', error);
    throw error;
  }
};

/**
 * Service function to delete a specific notification.
 * @param {number} id The ID of the notification to delete.
 * @returns {Promise} A promise that resolves when the notification is deleted.
 */
export const deleteNotification = async (id) => {
  try {
    const response = await api.delete(`${NOTIFICATIONS_URL}/${id}`);
    return response.data;
  } catch (error) {
    console.error(`Error in deleteNotification with id ${id}:`, error);
    throw error;
  }
};

/**
 * Service function to get the count of unread notifications for the current user.
 * @returns {Promise} A promise that resolves to the count of unread notifications.
 */
export const getUnreadNotificationsCount = async () => {
  try {
    const response = await api.get(`${NOTIFICATIONS_URL}/unread-count`);
    return response.data;
  } catch (error) {
    console.error('Error in getUnreadNotificationsCount:', error);
    throw error;
  }
};